abcclassroom/tests/test_git.py

```diff
@@ -1,28 +1,206 @@
+    # via a mock object too...
+    return sys.stdout.write(
+        "Hi username! You've successfully authenticated \n"
+    )
+
+
+# TODO - do we need this??
+def mock_call_git():
+    """A basic mock that returns a RuntimeError"""
+
+    # TODO: can i write to stderr as a return value perhaps? test this
+    # sys.stderr.write("Some git error here - test this")
+    raise RuntimeError("Some git error here - test this")
+
+
+@pytest.fixture()
+def example_student_repo(tmpdir):
+    """A fixture that creates an example student repo.
+    This can be run in a temp dir to avoid messing with a dev's local
+    environment. This allows us to test init and other repo functions
+    without needing to create files in our tests each time."""
+
+    repo_dir = Path(tmpdir, "assignment-1", "course-test-student")
+    repo_dir.mkdir(parents=True, exist_ok=True)
+    a_file = Path(repo_dir, "README.md")
+    a_file.write_text("Some text in the readme")
+
+    return repo_dir
+
+
+# TODO: add an  actual check here for the standard out
+def test_check_git_ssh_pass():
+    """When ssh is setup correctly, the check should run and pass with no
+    output."""
+    with mock.patch("subprocess.run"):
+        # Skip actually running the subprocess call and return the expected
+        # output
+        subprocess.run.side_effect = subprocess.CalledProcessError(
+            returncode=1, cmd="", stderr="Hi username"
+        )
+        # Here this should just run and pass
+        # I don't think we need an assert unless we add a message for
+        # successful check to the function. Then we get a return output that
+        # can be tested
+        abcgit.check_git_ssh()
+
+
+def test_check_git_ssh_warning(capsys):
+    """Test what happens when ssh is setup but user hasn't logged in"""
+    with mock.patch("subprocess.run"):
+        # Skip actually running the subprocess call and return the expected
+        # output. We always trigger a CalledProcessError in this function
+        subprocess.run.side_effect = subprocess.CalledProcessError(
+            returncode=1, cmd="", stderr="Warning: Permanently"
+        )
+
+        abcgit.check_git_ssh()
+        captured_output = capsys.readouterr().out.splitlines()
+        assert captured_output[0].startswith("Warning:")
+
+
+def test_check_git_ssh_error(capsys):
+    """Test what happens when ssh is not setup - should raise RuntimeError"""
+    with mock.patch("subprocess.run"):
+        # Skip running subprocess call, return the expected error
+        subprocess.run.side_effect = subprocess.CalledProcessError(
+            returncode=1, cmd="", stderr="Encountered this error"
+        )
+        with pytest.raises(RuntimeError):
+            abcgit.check_git_ssh()
+        captured_output = capsys.readouterr().out.splitlines()
+        assert captured_output[0].startswith("Encountered this error")
+
+
+# NOTE: Updated to run git status so it works universally
```

Comment:
@kcranston there is a bigger question here about how much we mock. Right now the big stumbling block is ssh setup for `git clone` but here i mean git status is easy to just run. Because you run git commit below and it's fine i may update this test to avoid any mocks.